src/fabric/src/fabric2_fdb.erl

```diff
@@ -346,12 +351,97 @@ delete(#{} = Db) ->
-    erlfdb:clear_range_startswith(Tx, DbPrefix),
+    DoRecovery = fabric2_util:do_recovery(),
+    case DoRecovery of
+        true ->
+            Timestamp = list_to_binary(fabric2_util:iso8601_timestamp()),
+            DeletedDbKey = erlfdb_tuple:pack({?DELETED_DBS, DbName, Timestamp},
+                LayerPrefix),
+            case erlfdb:wait(erlfdb:get(Tx, DeletedDbKey)) of
+                not_found ->
+                    erlfdb:set(Tx, DeletedDbKey, DbPrefix),
+                    erlfdb:clear(Tx, DbKey);
+                _Val ->
+                    erlang:error({deleted_database_exists, DbName})
+            end;
+        false ->
+            erlfdb:clear(Tx, DbKey),
+            erlfdb:clear_range_startswith(Tx, DbPrefix)
+    end,
     bump_metadata_version(Tx),
     ok.
 
 
+deleted_dbs_info(#{} = Db0) ->
+    #{
+        name := DbName,
+        tx := Tx,
+        layer_prefix := LayerPrefix
+    } = ensure_current(Db0, false),
+
+    DeletedDbKey =  erlfdb_tuple:pack({?DELETED_DBS, DbName}, LayerPrefix),
+    DeletedDbs = erlfdb:wait(erlfdb:get_range_startswith(Tx, DeletedDbKey)),
+    lists:foldl(fun({DbKey, DbPrefix}, Acc) ->
+        DBInfo = get_info_wait(get_info_future(Tx, DbPrefix)),
+        {?DELETED_DBS, DbName, DeletedTS} =
+            erlfdb_tuple:unpack(DbKey, LayerPrefix),
+        [{DeletedTS, DBInfo}| Acc]
+    end, [], DeletedDbs).
+
+
+undelete(#{} = Db0, TgtDbName, TimeStamp) ->
+    #{
+        name := DbName,
+        tx := Tx,
+        layer_prefix := LayerPrefix
+    } = ensure_current(Db0, false),
+    DbKey = erlfdb_tuple:pack({?ALL_DBS, TgtDbName}, LayerPrefix),
+    case erlfdb:wait(erlfdb:get(Tx, DbKey)) of
+        Bin when is_binary(Bin) ->
+            {error, file_exists};
+        not_found ->
+            DeletedDbTupleKey = {
+                ?DELETED_DBS,
+                DbName,
+                TimeStamp
+            },
+            DeleteDbKey =  erlfdb_tuple:pack(DeletedDbTupleKey, LayerPrefix),
+            case erlfdb:wait(erlfdb:get(Tx, DeleteDbKey)) of
+                not_found ->
+                    erlang:error({not_found, invalid_timestamp});
+                DbPrefix ->
+                    erlfdb:set(Tx, DbKey, DbPrefix),
+                    erlfdb:clear(Tx, DeleteDbKey),
+                    bump_metadata_version(Tx),
+                    ok
+            end
+    end.
+
+
+delete_deleted(#{} = Db0, TimeStamp) ->
+    #{
+        name := DbName,
+        tx := Tx,
+        layer_prefix := LayerPrefix
+    } = ensure_current(Db0),
+
+    DeletedDbTupleKey = {
+        ?DELETED_DBS,
+        DbName,
+        TimeStamp
+    },
+    DeleteDbKey =  erlfdb_tuple:pack(DeletedDbTupleKey, LayerPrefix),
+    case erlfdb:wait(erlfdb:get(Tx, DeleteDbKey)) of
+        not_found ->
+            erlang:error({not_found});
+        DbPrefix ->
+            erlfdb:clear(Tx, DeleteDbKey),
+            erlfdb:clear_range_startswith(Tx, DbPrefix),
+            bump_metadata_version(Tx)
```

Comment:
I don't think we need to bump the metadata version here since we're not affecting `_all_dbs`? What do you think, @nickva?

Reply:
I think we're ok not bumping it here if we guarantee (let's have a test for it) that any soft deleted db handle will reopen. For that we need to bump the db version when we soft-delete and ensure that to hard deletion (snapshot removal) always goes through the soft-deletion path first.

Reply:
Ooooh, good call on db versions.

We should call `bump_db_version` after soft deletion and undeletion so that any inflight requests will fail properly.

I don't think hard deletion or snapshot deletion require that check since clearing the key would create a write conflict, no need to change it just to delete it.

Reply:
Thinking more about tests I think we'd want at least for these scenarios:

Cases where old db handle can't read the data anymore:
  - db soft deleted
  - db hard deleted
  - db soft deleted and re-created
  - db soft deleted and undeleted
  - db hard deleted and re-created 
  
Indexing stops when:
   - db soft deleted
   - db hard deleted
   - db soft deleted and re-created
   - db soft deleted and undeleted
   - db hard deleted and re-created

Obviously indexing will stop if the first set of tests pass, however we want to make sure the way indexing detects that db is now deleted hasn't changed and it reacts the same as before. Look for db re-created or rename tests in recent commits and I think we could modify those to take soft deletion into consideration.

If we are using timestamps with 1 second resolution also make sure we handle cases were we soft-delete/undelete multiple times in the same second, and also tests where we do it across a few different seconds.

Reply:
Actually I was wrong about the `db soft deleted and undeleted`.  Thanks, @jiangphcn for pointing it out. It seems in that case the old handle will be usable because:

1) we bumped the db version only
2) old handle would notice metadata and db version being out of date
3) it would automatically re-open
4) it would continue to be used

We could assert that in the test perhaps then. Good catch @jiangphcn !

Reply:
Great suggestion @nickva on test case. I added it in https://github.com/apache/couchdb/pull/2666/commits/b1d2b08afb80ca9f8d8f4b9ca37ad65cd54b3352#diff-5c0948df6382504d34bd12ee0d4485d0R197-R250.

Reply:
Oh, for after creation. For that "use after deletion" we should make sure that between deletion and undeletion that the handle isn't usable as well.